sound/soc/sof/topology.c

```diff
@@ -1498,7 +1498,8 @@ static int sof_widget_ready(struct snd_soc_component *scomp, int index,
 	case snd_soc_dapm_kcontrol:
 	default:
 		dev_dbg(scomp->dev, "widget type %d name %s not handled\n", swidget->id, tw->name);
-		break;
+		kfree(swidget);
+		return 0;
```

Comment:
For the record:
I think the proper fixup would be:
```diff
diff --git a/sound/soc/sof/ipc3-topology.c b/sound/soc/sof/ipc3-topology.c
index 7870260c6f3e..88b9b9507231 100644
--- a/sound/soc/sof/ipc3-topology.c
+++ b/sound/soc/sof/ipc3-topology.c
@@ -2317,7 +2317,8 @@ static int sof_ipc3_tear_down_all_pipelines(struct snd_sof_dev *sdev, bool verif
 		if (!verify && !swidget->dynamic_pipeline_widget &&
 		    SOF_FW_VER(v->major, v->minor, v->micro) < SOF_FW_VER(2, 2, 0)) {
 			swidget->use_count = 0;
-			swidget->spipe->complete = 0;
+			if (swidget->spipe)
+				swidget->spipe->complete = 0;
 			continue;
 		}
 ```
Because the reason for the NULL pointer dereference is the fact that `sof_ipc3_tear_down_all_pipelines()` iterates through the global swidget list and the "ASoC: SOF: Introduce struct snd_sof_pipeline" move the `complete` flag under `spipe` which can be NULL if the swidget is not part of any path (virtual/not handled widgets).

This change would be better on it's own to be backportable when sending upstream.